### PR TITLE
Make Security Agent run as a service on Windows

### PR DESCRIPTION
**Pull request by ghost**
### What does this PR do?

Make security agent run as a service on Windows

### Motivation

To use FIM, security agent must run as SYSTEM, as a Windows service.

### Additional Notes

This commit split main.go to a NIX version and a Windows version. Main logic in app.go is changed to split the start and stop logic into RunAgent and StopAgent function. The old deferred function calls are moved to StopAgent.

This commit doesn't add the agent to installer package. It will be addressed in other PRs.

### Describe your test plan

Test on Windows by doing a build, copy security agent to Windows test box, run

New-Service -Name "datadog-security-agent" -StartupType "Manual" -BinaryPathName "C:\Program Files\Datadog\Datadog Agent\bin\agent\security-agent.exe"

### CODE REVIEW

**Review by derekwbrown — COMMENTED** [review]
the `ms000409.bin` is generated and shouldn't be checked in.

File: cmd/security-agent/windows_resources/security-agent.rc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+BEGIN
+    BLOCK "StringFileInfo"
+    BEGIN
+        BLOCK "040904b0"
+        BEGIN
+            VALUE "CompanyName", "Datadog, Inc."
+            VALUE "FileDescription", "Datadog Security Agent"
+            VALUE "FileVersion", FILE_VERSION_STRING
+            VALUE "InternalName", "security-agent"
+            VALUE "LegalCopyright", "Copyright (C) 2017"
```
Comment (derekwbrown) [inline cmd/security-agent/windows_resources/security-agent.rc:38]: Probably want to update this ;)

File: cmd/security-agent/windows_resources/security-agent-msg.mc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    English=0x409:MSG00409
+)
+
+
+;// Messages
+MessageId=1
+SymbolicName=MSG_WARN_REGCONFIG_FAILED
+Severity=Warning
+Language=English
+Failed to import config items from registry.  The error was %1
```
Comment (derekwbrown) [inline cmd/security-agent/windows_resources/security-agent-msg.mc:14]: Is this message used?

File: cmd/security-agent/windows_resources/security-agent-msg.mc [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+.
+
+MessageId=2
+SymbolicName=MSG_WARN_CONFIGUPGRADE_FAILED
+Severity=Warning
+Language=English
+Failed to upgrade configuration.  The error was %1.
```
Comment (derekwbrown) [inline cmd/security-agent/windows_resources/security-agent-msg.mc:21]: same with this

File: cmd/security-agent/windows_resources/security-agent.rc (hunk 1 above)
Comment (mikezhu-dd) [reply cmd/security-agent/windows_resources/security-agent.rc]: Updated to 2021

File: cmd/security-agent/windows_resources/security-agent-msg.mc (hunk 3 above)
Comment (mikezhu-dd) [reply cmd/security-agent/windows_resources/security-agent-msg.mc]: Removed

**Comment by mikezhu-dd** [comment]
Integration tests failed with my last commit. I checked the error and other PRs, all PRs failed at the same integration (dogstatsd) and it most likely caused by other commit.

File: cmd/security-agent/main_windows.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+)
+
+var (
+	elog           debug.Log
+	defaultLogFile = "c:\\programdata\\datadog\\logs\\security-agent.log"
+
+	// DefaultConfPath points to the folder containing datadog.yaml
+	DefaultConfPath = "c:\\programdata\\datadog"
+
+	enabledVals = map[string]bool{"yes": true, "true": true, "1": true,
```
Comment (derekwbrown) [inline cmd/security-agent/main_windows.go:42]: Not used?

File: tasks/security_agent.py [hunk 5]
```diff
@@ -57,6 +59,27 @@ def build(
     """
     ldflags, gcflags, env = get_build_flags(ctx, major_version=major_version, python_runtimes='3')
 
+    # generate windows resources
+    if sys.platform == 'win32':
+        windres_target = "pe-x86-64"
+        if arch == "x86":
```
Comment (derekwbrown) [inline tasks/security_agent.py:65]: I can't imagine we'd ever support this on x86.
Perhaps explicitly disable x86?

**Review by derekwbrown — APPROVED** [review]
Neither of my comments were significant... approved, and you can do what you want with the comments.

File: cmd/security-agent/app/app.go [hunk 6]
```diff
@@ -276,29 +296,66 @@ func start(cmd *cobra.Command, args []string) error {
...
+	}
+
+	if srv != nil {
+		srv.Stop()
 	}
 
 	log.Info("See ya!")
-	return nil
+	log.Flush()
+	return
```
Comment (safchain) [inline cmd/security-agent/app/app.go:360]: nit: you can remove this line

File: cmd/security-agent/app/app.go [hunk 7]
```diff
@@ -276,29 +296,66 @@ func start(cmd *cobra.Command, args []string) error {
...
-	<-signalCh
+	for signo := range signalCh {
+		switch signo {
+		case syscall.SIGPIPE:
+			// By default systemd redirects the stdout to journald. When journald is stopped or crashes we receive a SIGPIPE signal.
+			// Go ignores SIGPIPE signals unless it is when stdout or stdout is closed, in this case the agent is stopped.
+			// We never want dogstatsd to stop upon receiving SIGPIPE, so we intercept the SIGPIPE signals and just discard them.
+		default:
+			log.Infof("Received signal '%s', shutting down...", signo)
+			stopCh <- struct{}{}
```
Comment (safchain) [inline cmd/security-agent/app/app.go:329]: What about closing it ?

File: cmd/security-agent/main_windows.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		return
+	}
+	elog.Info(0x40000003, ServiceName)
+
+	changes <- svc.Status{State: svc.Running, Accepts: cmdsAccepted}
+
+loop:
+	for {
+		select {
+		case c := <-r:
```
Comment (safchain) [inline cmd/security-agent/main_windows.go:123]: as you have only one `case` I think you can do `for c := range r {`

File: cmd/security-agent/app/app.go (hunk 7 above)
Comment (mikezhu-dd) [reply cmd/security-agent/app/app.go:329]: Fixed in func start

File: tasks/security_agent.py (hunk 5 above)
Comment (mikezhu-dd) [reply tasks/security_agent.py:65]: Will do that later.

**Pull request merged**